File: docs/domain_token.py
"""Reference implementation of domain registration token"""

import base64
import hashlib
import hmac
import sys
import time
import typing
import uuid

PERSONALITY = b"register domain"

# namespace UUID for IDMSVC
# uuid.uuid5(uuid.NAMESPACE_URL, "https://console.redhat.com/api/idmsvc")
NS_IDMSVC = uuid.UUID("2978cc95-31c8-503d-ba8f-581911b6bea0")

# Python 3.6 does not have time.time_ns()
if sys.version_info < (3, 7):

    def _time_ns():
        return int(time.time() * 1_000_000_000)

else:
    _time_ns = time.time_ns


def token_domain_id(token: str) -> uuid.UUID:
    """Get domain UUID from a token string"""
    return uuid.uuid5(NS_IDMSVC, token)


def generate_token(
    key: bytes, domain_type: str, org_id: str, *, validity: int = 2 * 60 * 60
) -> typing.Tuple[str, int]:
    """Generate a domain registration token"""
    expires = int(_time_ns() + (validity * 1_000_000_000))
    token = _generate_token_ns(key, domain_type, org_id, expires)
    return token, expires


def _generate_token_ns(key: bytes, domain_type: str, org_id: str, expires: int) -> str:
    payload_bytes = expires.to_bytes(8, "big")
    payload_b64 = _b64encode(payload_bytes)
    sig = _mac_digest(key, domain_type, org_id, payload_bytes)
    sig_b64 = _b64encode(sig)
    return f"{payload_b64}.{sig_b64}"


def validate_token(key: bytes, domain_type: str, org_id: str, token: str) -> uuid.UUID:
    """Validate a domain registration token"""
    expires = _validate_token_sig(key, domain_type, org_id, token)
    if _time_ns() > expires:
        raise ValueError("token expired")
    return token_domain_id(token)


def _validate_token_sig(key: bytes, domain_type: str, org_id: str, token: str) -> int:
    payload_b64, sig_b64 = token.split(".")
    payload_bytes = _b64decode(payload_b64)
    sig = _b64decode(sig_b64)
    digest = _mac_digest(key, domain_type, org_id, payload_bytes)
    if not hmac.compare_digest(sig, digest):
        raise ValueError("Invalid signature")
    return int.from_bytes(payload_bytes, "big")


def _b64encode(payload: bytes) -> str:
    """URL-safe base64 encoding without padding"""
    return base64.urlsafe_b64encode(payload).rstrip(b"=").decode("ascii")


def _b64decode(payload: str) -> bytes:
    """URL-safe base64 decoding without padding"""
    rem = len(payload) % 4
    if rem == 2:
        payload += "=="
    elif rem == 3:
        payload += "="
    return base64.urlsafe_b64decode(payload.encode("utf-8"))


def _mac_digest(
    key: bytes, domain_type: str, org_id: str, payload_bytes: bytes
) -> bytes:
    """Create MAC from org_id and payload"""
    mac = hmac.HMAC(key, digestmod=hashlib.sha256)
    mac.update(PERSONALITY)
    mac.update(domain_type.encode("utf-8"))
    mac.update(org_id.encode("utf-8"))
    mac.update(payload_bytes)
    return mac.digest()


def test() -> None:
    KEY = b"secretkey"
    ORG_ID = "123456"
    DOMAIN_TYPE = "rhel-idm"
    token, expires = generate_token(KEY, DOMAIN_TYPE, ORG_ID)
    print(token, expires)
    domain_id = validate_token(KEY, DOMAIN_TYPE, ORG_ID, token)
    print(domain_id)
    # validate_token(KEY, "123123", token)


if __name__ == "__main__":
    test()

File: internal/infrastructure/token/domain_token/domain_token.go
/* Domain Registration Token
 *
 * see docs/domain-token.md for more information
 */
package domain_token

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/binary"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
)

type DomainRegistrationToken string

var (
	// uuid.NewSHA1(uuid.NameSpaceURL, []byte("https://console.redhat.com/api/idmsvc"))
	NamespaceIDMSVC           = uuid.MustParse("2978cc95-31c8-503d-ba8f-581911b6bea0")
	RegisterDomainPersonality = []byte("register domain")
)

// Derive domain id from token string
func TokenDomainId(token DomainRegistrationToken) uuid.UUID {
	return uuid.NewSHA1(NamespaceIDMSVC, []byte(token))
}

// Create a new domain registration token
// The token is signed by *key*, bound to *orgID*, and validate until
// now + validity duration.
func NewDomainRegistrationToken(
	key []byte, domainType string, orgID string, validity time.Duration,
) (token DomainRegistrationToken, expireNS uint64, err error) {
	expireNS = uint64(time.Now().UnixNano() + validity.Nanoseconds())
	tok, err := newDomainRegistrationTokenAt(key, domainType, orgID, uint64(expireNS))
	return tok, expireNS, err
}

// Create a domain registration token that expires at *expireNS* nanoseconds
// after Unix epoch.
func newDomainRegistrationTokenAt(
	key []byte, domainType string, orgID string, expireNS uint64,
) (token DomainRegistrationToken, err error) {
	payload_bytes := make([]byte, 8)
	binary.BigEndian.PutUint64(payload_bytes, expireNS)
	payload_b64 := base64.RawURLEncoding.EncodeToString(payload_bytes)

	sig := mac_digest(key, domainType, orgID, payload_bytes)
	sig_b64 := base64.RawURLEncoding.EncodeToString(sig)

	return DomainRegistrationToken(fmt.Sprintf("%s.%s", payload_b64, sig_b64)), nil
}

// Verify signature, *orgID* binding, and expiration time stamp of a token.
// Returns the domain UUID on success.
func VerifyDomainRegistrationToken(
	key []byte, domainType string, orgID string, token DomainRegistrationToken,
) (domainId uuid.UUID, err error) {
	var expireNS uint64
	if expireNS, err = parseDomainRegistrationToken(key, domainType, orgID, token); err != nil {
		return uuid.Nil, err
	}
	var now uint64 = uint64(time.Now().UnixNano())
	if now > expireNS {
		return uuid.Nil, fmt.Errorf("Token has expired: %d > %d", now, expireNS)
	}
	return TokenDomainId(token), nil
}

// Parse and check signature of token
func parseDomainRegistrationToken(
	key []byte, domainType string, orgID string, token DomainRegistrationToken,
) (expireNS uint64, err error) {
	var (
		payload_bytes []byte
		sig           []byte
	)
	if len(token) > 100 {
		return 0, fmt.Errorf("Token length exceeded")
	}
	parts := strings.Split(string(token), ".")
	if len(parts) != 2 {
		return 0, fmt.Errorf("Invalid token")
	}
	if payload_bytes, err = base64.RawURLEncoding.DecodeString(parts[0]); err != nil {
		return 0, err
	}
	if sig, err = base64.RawURLEncoding.DecodeString(parts[1]); err != nil {
		return 0, err
	}
	expected_sig := mac_digest(key, domainType, orgID, payload_bytes)
	if !hmac.Equal(sig, expected_sig) {
		return 0, fmt.Errorf("Signature mismatch")
	}
	return binary.BigEndian.Uint64(payload_bytes), nil
}

// Calculate keyed MAC digest from orgID and payload
func mac_digest(key []byte, domainType string, orgID string, payload []byte) []byte {
	mac := hmac.New(sha256.New, key)
	// Hash.Write() never returns an error
	mac.Write(RegisterDomainPersonality)
	mac.Write([]byte(domainType))
	mac.Write([]byte(orgID))
	mac.Write(payload)
	return mac.Sum(nil)
}
